src/server/pfs/server/driver.go

```diff
@@ -301,6 +301,462 @@ func (d *driver) getAccessLevel(pachClient *client.APIClient, repo *pfs.Repo) (a
+		// i.e. union(branch, branch.Provenance) = union(branch, branch.DirectProvenance, branch.DirectProvenance.Provenance)
+		direct := bi.DirectProvenance
+		union := []*pfs.Branch{bi.Branch}
+		for _, directProvenance := range direct {
+			directProvenanceInfo := branchInfos[key(directProvenance.Repo.Name, directProvenance.Name)]
+			union = append(union, directProvenance)
+			union = append(union, directProvenanceInfo.Provenance...)
+		}
+
+		if !equalBranches(append(bi.Provenance, bi.Branch), union) {
+			return ErrBranchProvenanceTransitivity{
+				BranchInfo:     bi,
+				FullProvenance: union,
+			}
+		}
+
+		// 	if there is a HEAD commit
+		if bi.Head != nil {
+			// we expect the branch's provenance to equal the HEAD commit's provenance
+			// i.e branch.Provenance contains the branch provBranch and provBranch.Head != nil implies branch.Head.Provenance contains provBranch.Head
+			// =>
+			for _, provBranch := range bi.Provenance {
+				provBranchInfo, ok := branchInfos[key(provBranch.Repo.Name, provBranch.Name)]
+				if !ok {
+					return ErrBranchInfoNotFound{Branch: provBranch}
+				}
+				if provBranchInfo.Head != nil {
+					// in this case, the headCommit Provenance should contain provBranch.Head
+					headCommitInfo, ok := commitInfos[key(bi.Head.Repo.Name, bi.Head.ID)]
+					if !ok {
+						return ErrCommitInfoNotFound{
+							Location: "head commit provenance (=>)",
+							Commit:   bi.Head,
+						}
+					}
+					contains := false
+					for _, headProv := range headCommitInfo.Provenance {
+						if provBranchInfo.Head.Repo.Name == headProv.Commit.Repo.Name &&
+							provBranchInfo.Branch.Repo.Name == headProv.Branch.Repo.Name &&
+							provBranchInfo.Name == headProv.Branch.Name &&
+							provBranchInfo.Head.ID == headProv.Commit.ID {
+							contains = true
+						}
+					}
+					if !contains {
+						return ErrHeadProvenanceInconsistentWithBranch{
+							BranchInfo:     bi,
+							ProvBranchInfo: provBranchInfo,
+							HeadCommitInfo: headCommitInfo,
+						}
+					}
+				}
+			}
+			// <= (Not necessarily true)
+			// headCommitInfo, ok := commitInfos[key(bi.Head.Repo.Name, bi.Head.ID)]
+			// if !ok {
+			// 	return ErrCommitInfoNotFound{
+			// 		Location: "head commit provenance (<=)",
+			// 		Commit:   bi.Head,
+			// 	}
+			// }
+			// for _, headProv := range headCommitInfo.Provenance {
+			// 	contains := false
+			// 	for _, provBranch := range bi.Provenance {
+			// 		provBranchInfo, ok := branchInfos[key(provBranch.Repo.Name, provBranch.Name)]
+			// 		if !ok {
+			// 			return ErrBranchInfoNotFound{Branch: provBranch}
+			// 		}
+			// 		if headProv.Branch.Repo.Name == provBranchInfo.Branch.Repo.Name &&
+			// 			headProv.Branch.Name == provBranchInfo.Branch.Name {
+			// 			if provBranchInfo.Head == nil {
+			// 				return fmt.Errorf("")
+			// 			}
+			// 			contains = true
+			// 		}
+			// 	}
+			// 	if !contains {
+			// 		return fmt.Errorf("")
+			// 	}
+			// }
```

Comment:
This should probably either be removed or uncommented, it'll bitrot if left in this state.

Reply:
Yeah, I'll remove it. 

I think this invariant as is doesn't work, but if something like this is true, it would probably be an adjoint functor to `Head`

Reply:
There don't seem to be any natural functors in the commit to branch direction, so no adjoints exist. So that makes me believe that having anything like this code block would be an unnatural invariant.